src/potions.ts

```diff
@@ -417,3 +430,134 @@ export function bathroomFinance(embezzlers: number): void {
+  meatBonusPerTurn: number; // meat% bonus per turn
+  cappedMeatBonus: number;
+
+  constructor(
+    potion: Item,
+    softcap: number,
+    meatBonusPerTurn: number,
+    duration?: number,
+    effect?: Effect
+  ) {
+    this.potion = potion;
+    this.effect = effect ?? effectModifier(potion, "Effect");
+    this.duration = duration ?? numericModifier(potion, "Effect Duration");
+    this.softcap = softcap;
+    this.meatBonusPerTurn = meatBonusPerTurn;
+    this.cappedMeatBonus = softcap * meatBonusPerTurn;
+  }
+
+  use(quantity: number): boolean {
+    return use(quantity, this.potion);
+  }
+
+  price(historical: boolean): number {
+    // If asked for historical, and age < 14 days, use historical.
+    // If potion is not tradeable, use retrievePrice instead
+    return this.potion.tradeable
+      ? historical && historicalAge(this.potion) < 14
+        ? historicalPrice(this.potion)
+        : mallPrice(this.potion)
+      : retrievePrice(this.potion);
+  }
+
+  getOptimalNumberToUse(yachtzees: number, embezzlers: number): number {
+    const barfTurns = Math.max(0, estimatedTurns() - yachtzees - embezzlers);
+
+    const potionAmountsToConsider = [] as number[];
+    for (const fn of [Math.floor, Math.ceil]) {
+      for (const sc of [0, this.softcap]) {
+        for (const em of [0, embezzlers]) {
+          for (const bt of [0, barfTurns]) {
+            potionAmountsToConsider.push(fn((yachtzees + em + bt + sc) / this.duration));
+          }
+        }
+      }
+    }
+
+    const profits = [] as number[];
+    const profitsMap = new Map(
+      potionAmountsToConsider.map((n) => {
+        const profit = this.valueNPotions(n, yachtzees, embezzlers, barfTurns);
+        profits.push(profit);
+        return [profit, n];
+      })
+    );
+
+    const bestProfits = profits.reduce((a, b) => (a > b ? a : b));
+    if (bestProfits > 0) {
+      const nPotionsToUse =
+        potionAmountsToConsider[profitsMap.get(bestProfits) ?? 0] -
+        Math.floor(haveEffect(this.effect) / this.duration);
+      return Math.max(nPotionsToUse, 0);
+    }
+    return 0;
+  }
+
+  valueNPotions(n: number, yachtzees: number, embezzlers: number, barfTurns: number): number {
+    const yachtzeeValue = 2000;
+    const embezzlerValue = baseMeat + 750;
+    const barfValue = (baseMeat * turnsToNC) / 30;
+
+    const totalCosts = n * this.price(false);
+    const totalDuration = n * this.duration;
+    let cappedDuration = Math.max(0, totalDuration - this.softcap + 1);
+    let decayDuration = Math.min(totalDuration, this.softcap - 1);
+    let totalValue = 0;
+    const turnTypes = [
+      [yachtzees, yachtzeeValue],
+      [embezzlers, embezzlerValue],
+      [barfTurns, barfValue],
+    ];
+
+    // eslint-disable-next-line prefer-const
+    for (let [turns, value] of turnTypes) {
+      if (cappedDuration >= turns) {
+        totalValue += (value * turns * this.cappedMeatBonus) / 100;
+        cappedDuration -= turns;
```

Comment:
It might (this is really a "might", not a gently phrased "do this") read better if instead of doing `if...else`, you did
```
if (condition) {
  doTheThing();
  continue;
}
more code here
```
continue works a little like an early return in a function; it'll just move on to the next phase of the loop

Reply:
It could also be nicer to slice the stuff inside this function out into a loop, and then you can literally use `return`

Reply:
I don't know if I agree with the `continue` comment. I'd use `continue` and early returns when the condition I'm considering is really specific/niche, but this feels more like a red pill/blue pill thing. It feels more like either/or, which I find it easier to read as an if/else.

I don't like this current structure either, but I do need to capture `cappedDuration`, `decayDuration` and `totalValue` in this scope (technically I don't really need to capture `totalValue`). I'm not sure if I'm following what you have in mind when you mention slicing the stuff out.

Reply:
Really these should probably be two functions that are selected by this if statement. You'll need to work around decrementing the various values in doing so, but it'll be better code.

Reply:
Aight, no more if/else thanks to the power of maths (but mostly thanks to more braincells being available today). Hopefully this is much better code